File: test_javaPrj_01/src/programmers/lv1/공원산책.java
package programmers.lv1;

import java.util.Arrays;

public class 공원산책 {

    private 공원산책(){}

    public void Solution() throws Exception{
        String[] park = {"OOOXO", "OOOOO", "OOOSO", "OOOXO", "OOOXO", "OOOOO"};
        String[] routes = {"S 1"};

        String parkString = Arrays.toString(park).replaceAll("[^OSX]", "");
        
        int parkWidth = park[0].length();
        int parkHeight = park.length;

        String[] parkRotated = new String[parkWidth];
        Arrays.fill(parkRotated, "");
        for(int i = 0; i < parkString.length(); i++){
            parkRotated[i%parkWidth] += parkString.charAt(i);            
        }

        System.out.println(Arrays.toString(park));
        System.out.println(Arrays.toString(parkRotated));

        int startIdx = parkString.indexOf("S");
        int yStart = startIdx/park[0].length();
        int xStart = startIdx%park[0].length();
        Dog dog = new Dog(xStart, yStart);
        
        for(String route : routes){
            String direction = route.replaceAll("[0-9 ]", "");
            int distance = Integer.valueOf(route.replaceAll("[^0-9]", ""));

            if("E".equals(direction)){
                System.out.println(dog.getX()+distance > parkWidth-1);
                System.out.println("E : "+park[dog.getY()].substring(dog.getX()+1, dog.getX()+distance+1));
                if(dog.getX()+distance > parkWidth-1 || park[dog.getY()].substring(dog.getX()+1, dog.getX()+distance+1).indexOf("X") > -1){
                    continue;
                };
                dog.moveToEast(distance);
            }else if("W".equals(direction)){
                System.out.println(dog.getX()-distance < 0);
                System.out.println("W : "+park[dog.getY()].substring(dog.getX()-distance, dog.getX()));
                if(dog.getX()-distance < 0 || park[dog.getY()].substring(dog.getX()-distance, dog.getX()).indexOf("X") > -1){
                    continue;
                };
                dog.moveToWest(distance);
            }else if("N".equals(direction)){
                System.out.println(dog.getY()-distance < 0);
                System.out.println("N : "+parkRotated[dog.getX()].substring(dog.getY()-distance, dog.getY()));
                if(dog.getY()-distance < 0 || parkRotated[dog.getX()].substring(dog.getY()-distance, dog.getY()).indexOf("X") > -1 ){
                    continue;
                }
                dog.moveToNorth(distance);
            }else if("S".equals(direction)){
                System.out.println(dog.getY()+distance > parkHeight-1);
                System.out.println("S : "+parkRotated[dog.getX()].substring(dog.getY()+1, dog.getY()+distance+1));
                if(dog.getY()+distance > parkHeight-1 || parkRotated[dog.getX()].substring(dog.getY(), dog.getY()+distance).indexOf("X") > -1 ){
                    continue;
                }
                dog.moveToSouth(distance);
            }
        }

        System.out.println("Dog Location : "+dog.getX()+" : "+dog.getY());

    }

    class Dog{
        private int x;
        private int y;

        private Dog(int x, int y){
            this.x = x;
            this.y = y;
        }

        public void moveToNorth(int distance){
            this.y = this.y-distance;
        }
        public void moveToSouth(int distance){
            this.y = this.y+distance;
        }
        public void moveToEast(int distance){
            this.x = this.x+distance;
        }
        public void moveToWest(int distance){
            this.x = this.x-distance;
        }

        public int getX() {
            return this.x;
        }

        public void setX(int x) {
            this.x = x;
        }

        public int getY() {
            return this.y;
        }

        public void setY(int y) {
            this.y = y;
        }

    }



    public static void main(String[] args) {
        
        try{
            new 공원산책().Solution();
        }catch(Exception e){
            e.printStackTrace();
        }


    }
    
}

File: test_javaPrj_01/src/programmers/lv1/공원산책.py



import re
class 공원산책:

    def __init__(self, park:list[str], routes:list[str]) -> None:

        # List.toString
        parkStr = ''.join(park)

        parkWidth = len(park[0])
        parkHeight = len(park)

        # None으로 채운 빈 방 + Generic 리스트 선언법
        parkRotated:list[str] = [None for _ in range(parkWidth)]
        
        for i in range(len(parkStr)):
            # 유사 삼항연산자 이용
            parkRotated[i%parkWidth] = parkStr[i] if parkRotated[i%parkWidth] is None else parkRotated[i%parkWidth] + parkStr[i]
        
        # String.charAt
        startIdx = parkStr.find('S')
        xStart = startIdx%parkWidth
        # 나누기 중 몫만 구하기(일반 나누기는 실수형)
        yStart = startIdx//parkWidth
        dog = Dog(xStart, yStart)

        for route in routes:

            # String replaceAll 정규식
            direction = re.sub('[0-9 ]', '', route)
            distance = int(re.sub('[^0-9]', '', route))

            # 값 비교 ==, 주소 비교 is
            if 'E' == direction:
 
                # 논리연산자(and, or), String subString indexOf
                if dog.x+distance > parkWidth-1 or park[dog.y][dog.x+1:dog.x+distance+1].find('X') > -1 :
                    continue

                dog.moveToEast(distance)

            elif 'W' == direction:
                
                if dog.x-distance < 0 or park[dog.y][dog.x-distance:dog.x].find('X') > -1:
                    continue
                    
                dog.moveToWest(distance)


            elif 'N' == direction:

                if dog.y-distance < 0 or parkRotated[dog.x][dog.y-distance:dog.y].find('X') > -1:
                    continue

                dog.moveToNorth(distance)

                    
            elif 'S' == direction:
                
                if dog.y+distance > parkHeight-1 or parkRotated[dog.x][dog.y+1:dog.y+distance+1].find('X') > -1:
                    continue

                dog.moveToSouth(distance)

        
        

class Dog:

    def __init__(self, x:int, y:int) -> None:     
        # 인스턴스 변수? 속성? 선언법   
        self.x = x
        self.y = y

    def moveToNorth(self, distance:int):
        self.y = self.y-distance

    def moveToSouth(self, distance:int):
        self.y = self.y+distance
    
    def moveToEast(self, distance:int):
        self.x = self.x+distance
    
    def moveToWest(self, distance:int):
        self.x = self.x-distance
        


공원산책(["OOOXO", "OOOOO", "OXOSX", "OOOOO", "OOOXO", "OOOOO"], ["S 2", "E 1"])